snuba/state/cache/redis/backend.py

```diff
@@ -0,0 +1,220 @@
+        # wait for a different client to finish the work. We have to pass the
+        # task creation parameters -- the timeout (execution deadline) and a
+        # new task identity just in case we are the first in line.
+        result = self.__script_get(
+            [result_key, wait_queue_key, task_ident_key], [timeout, uuid.uuid1().hex]
+        )
+
+        if result[0] == RESULT_VALUE:
+            # If we got a cache hit, this is easy -- we just return it.
+            logger.debug("Immediately returning result from cache hit.")
+            return self.__codec.decode(result[1])
+        elif result[0] == RESULT_EXECUTE:
+            # If we were the first in line, we need to execute the function.
+            # We'll also get back the task identity to use for sending
+            # notifications and approximately how long we have to run the
+            # function. (In practice, these should be the same values as what
+            # we provided earlier.)
+            task_ident = result[1].decode("utf-8")
+            task_timeout = int(result[2])
+            logger.debug(
+                "Executing task (%r) with %s second timeout...",
+                task_ident,
+                task_timeout,
+            )
+
+            argv = [task_ident, 60]
+            try:
+                # The task is run in a thread pool so that we can return
+                # control to the caller once the timeout is reached.
+                value = self.__executor.submit(function).result(task_timeout)
+                argv.extend(
+                    [self.__codec.encode(value), get_config("cache_expiry_sec", 1)]
+                )
+            except concurrent.futures.TimeoutError as error:
+                raise TimeoutError("timed out waiting for value") from error
+            finally:
+                # Regardless of whether the function succeeded or failed, we
+                # need to mark the task as completed. If there is no result
+                # value, other clients will know that we raised an exception.
+                logger.debug("Setting result and waking blocked clients...")
+                try:
+                    self.__script_set(
+                        [
+                            result_key,
+                            wait_queue_key,
+                            task_ident_key,
+                            build_notify_queue_key(task_ident),
+                        ],
+                        argv,
+                    )
+                except ResponseError:
+                    # An error response here indicates that we overran our
+                    # deadline, or there was some other issue when trying to
+                    # put the result value in the cache. This doesn't affect
+                    # _our_ evaluation of the task, so log it and move on.
+                    logger.warning("Error setting cache result!", exc_info=True)
```

Comment:
This will still get called if we encountered a timeout, as in this example:

```python
>>> try:
...     raise Exception('a')
... except Exception as e:
...     print('in except')
...     raise Exception('b') from e
... finally:
...     print('in finally')
... 
in except
in finally
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
Exception: a

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
Exception: b
```

That won't break anything, but isn't the most efficient thing — the only way that I could anticipate this not raising a `ResponseError` and being discarded is if our clock is running much faster than Redis. It'd probably be better to just skip this block entirely in that case.